src/libstd/sys/unix/fs.rs

```diff
@@ -132,6 +139,32 @@ impl FromInner<raw::mode_t> for FilePermissions {
 impl Iterator for ReadDir {
     type Item = io::Result<DirEntry>;
 
+    #[cfg(target_os = "sunos")]
```

Comment:
There is a `readdir_r` function, but it's non-standard (i.e. it doesn't comply to POSIX). And if you refer to the manpage for [`readdir(3C)`](https://illumos.org/man/3C/readdir) on Illumos, it says the following:

       It is safe to use readdir() in a threaded application, so long as only
       one thread reads from the directory stream at any given time. The
       readdir() function is generally preferred over the readdir_r() function.